library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java

```diff
@@ -357,24 +357,34 @@ private boolean updateLocked() {
     }
 
     private List<DownloadBatch> fetchBatches(Collection<DownloadInfo> downloads) {
-        List<DownloadBatch> batches = new ArrayList<>();
         Cursor batchesCursor = resolver.query(Downloads.Impl.BATCH_CONTENT_URI, null, null, null, null);
-        batches.clear();
+        List<DownloadBatch> batches = new ArrayList<>(batchesCursor.getCount());
+        List<Long> forDeletion = new ArrayList<>();
         try {
             int idColumn = batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches._ID);
+            int deleteIndex = batchesCursor.getColumnIndex(Downloads.Impl.Batches.COLUMN_DELETED);
+            int titleIndex = batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_TITLE);
+            int descriptionIndex = batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_DESCRIPTION);
+            int bigPictureUrlIndex = batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_BIG_PICTURE);
+            int statusIndex = batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_STATUS);
             int visibilityColumn = batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_VISIBILITY);
             while (batchesCursor.moveToNext()) {
                 long id = batchesCursor.getLong(idColumn);
 
-                String title = batchesCursor.getString(batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_TITLE));
-                String description = batchesCursor.getString(batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_DESCRIPTION));
-                String bigPictureUrl = batchesCursor.getString(batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_BIG_PICTURE));
-                int status = batchesCursor.getInt(batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches.COLUMN_STATUS));
+                if (batchesCursor.getInt(deleteIndex) == 1) {
+                    forDeletion.add(id);
+                    continue;
+                }
+
+                String title = batchesCursor.getString(titleIndex);
+                String description = batchesCursor.getString(descriptionIndex);
+                String bigPictureUrl = batchesCursor.getString(bigPictureUrlIndex);
+                int status = batchesCursor.getInt(statusIndex);
                 @NotificationVisibility.Value int visibility = batchesCursor.getInt(visibilityColumn);
 
                 BatchInfo batchInfo = new BatchInfo(title, description, bigPictureUrl, visibility);
 
-                List<DownloadInfo> batchDownloads = new ArrayList<>();
+                List<DownloadInfo> batchDownloads = new ArrayList<>(1);
```

Comment:
More *The More You Know*
Adding to an `ArrayList` that's full will double its capacity, unless the capacity is less than 6 in which case it will increase the capacity by 12: http://androidxref.com/5.1.0_r1/xref/libcore/luni/src/main/java/java/util/ArrayList.java#114